crates/wasi-common/src/snapshots/wasi_snapshot_preview1.rs

```diff
@@ -0,0 +1,1448 @@
+            Descriptor::VirtualFile(_virt) => {
+                unimplemented!("virtual filestat_set_times");
+            }
+            _ => path::filestat_set_times(resolved, flags, atim, mtim, fst_flags),
+        }
+    }
+
+    fn path_link(
+        &self,
+        old_fd: types::Fd,
+        old_flags: types::Lookupflags,
+        old_path: &GuestPtr<'_, str>,
+        new_fd: types::Fd,
+        new_path: &GuestPtr<'_, str>,
+    ) -> Result<()> {
+        trace!(
+            "path_link(old_fd={:?}, old_flags={:?}, old_path={:?}, new_fd={:?}, new_path={:?})",
+            old_fd,
+            old_flags,
+            old_path,
+            new_fd,
+            new_path,
+        );
+
+        // FIXME
+        // Much like in the case with `path_symlink`, we need to account for the
+        // fact where both `old_path` and `new_path` point the same memory location.
+        // Hence, separate scopes and `GuestBorrows` instances.
+        let resolved_old = {
+            let mut bc = GuestBorrows::new();
+            let old_as_str = old_path.as_raw(&mut bc)?;
+            let old_path = unsafe { &*old_as_str };
+
+            trace!("     | (old_path_ptr,old_path_len)='{}'", old_path);
+
+            let old_entry = unsafe { self.get_entry(old_fd)? };
+            path::get(
+                &old_entry,
+                types::Rights::PATH_LINK_SOURCE,
+                types::Rights::empty(),
+                types::Lookupflags::empty(),
+                old_path,
+                false,
+            )?
+        };
+        let resolved_new = {
+            let mut bc = GuestBorrows::new();
+            let new_as_str = new_path.as_raw(&mut bc)?;
+            let new_path = unsafe { &*new_as_str };
+
+            trace!("     | (new_path_ptr,new_path_len)='{}'", new_path);
+
+            let new_entry = unsafe { self.get_entry(new_fd)? };
+            path::get(
+                &new_entry,
+                types::Rights::PATH_LINK_TARGET,
+                types::Rights::empty(),
+                types::Lookupflags::empty(),
+                new_path,
+                false,
+            )?
+        };
+        path::link(resolved_old, resolved_new)
+    }
+
+    fn path_open(
+        &self,
+        dirfd: types::Fd,
+        dirflags: types::Lookupflags,
+        path: &GuestPtr<'_, str>,
+        oflags: types::Oflags,
+        fs_rights_base: types::Rights,
+        fs_rights_inheriting: types::Rights,
+        fdflags: types::Fdflags,
+    ) -> Result<types::Fd> {
+        trace!(
+            "path_open(dirfd={:?}, dirflags={}, path={:?}, oflags={}, fs_rights_base={}, fs_rights_inheriting={}, fdflags={}",
+            dirfd,
+            dirflags,
+            path,
+            oflags,
+            fs_rights_base,
+            fs_rights_inheriting,
+            fdflags,
+        );
+
+        // Extract path as &str.
+        let mut bc = GuestBorrows::new();
+        let path = unsafe { &*path.as_raw(&mut bc)? };
```

Comment:
That's correct, `as_raw` does the bounds checking. We need `unsafe` here since we're dereferencing a raw `*mut _` pointer, and in this particular case, we're obtaining `&str`. The signature of this function here is:

```rust
fn as_raw(...) -> Result<*mut _, GuestError>
```

 This is unsafe since we cannot really track the references here other than by using `GuestBorrows`. More on that [here](https://github.com/bytecodealliance/wasmtime/blob/master/crates/wiggle/crates/runtime/src/lib.rs#L401).